drud/ddev
# Basic port-test capabilities before trying to bring up router, fixes #126, fixes #393

## The Problem/Issue/Bug:

#126 explains: we need to test ports 80 and 443 before we let docker-compose try to bring up the router. If we find that somebody is already listening on them, we need to explain the problem to the user gracefully and provide resources to resolve the problem.

This also fixes #393 (Stopped site, double notifications of broken router)

## How this PR Solves The Problem:

* Stop the router before starting it again.
* Test our ports at that time
* If there's a problem stop then and explain to the user, giving resources.

## Manual Testing Instructions:

* Stop all sites (or manually use docker to stop the router)
* Occupy port 80 or 443 using a technique like `docker run -p 80:80 -p 443:443 --entrypoint=/bin/cat -it --rm busybox`  You can also start netcat or apache or nginx or many other ways.
* Run `ddev start` with or without, you should see correct behavior.

To test the fix for #393 
* Start one site
* Stop the router using docker `docker stop ddev-router`
* Use `ddev list` and `ddev describe` to look for the issue

This now has:
- [x]  Documentation (troubleshooting) links to help people to resolve their issue.
- [x]  An automated test to verify behavior.
- [x] A link to the troubleshooting docs when this problem happens.

## Automated Testing Overview:

Added TestRouterPortsCheck() which occupies port 80 using a docker container and then tests to see if we can start the router.

## Related Issue Link(s):

OP #126 

## Release/Deployment notes:
